include/zephyr/logging/log.h

```diff
@@ -89,13 +91,334 @@ extern "C" {
  */
 #define LOG_WRN_ONCE(...)					\
 	do {							\
-		static uint8_t __warned;			\
-		if (unlikely(__warned == 0)) {			\
+		static atomic_t __warned;			\
+		if (unlikely(atomic_cas(&__warned, 0, 1))) {	\
 			Z_LOG(LOG_LEVEL_WRN, __VA_ARGS__);	\
-			__warned = 1;				\
 		}						\
 	} while (0)
 
+/**
+ * @brief Core rate-limited logging macro
+ *
+ * @details Internal macro that provides rate-limited logging functionality.
+ * Uses atomic operations to ensure thread safety in multi-threaded environments.
+ * Only one thread can successfully log a message within the specified rate limit period.
+ *
+ * @param _level Log level (LOG_LEVEL_ERR, LOG_LEVEL_WRN, etc.)
+ * @param _rate_ms Minimum interval in milliseconds between log messages.
+ * @param _skipped_msg Message to show when messages are skipped.
+ * @param _log_func Logging function to call (Z_LOG or Z_LOG_HEXDUMP)
+ * @param ... Arguments to pass to the logging function.
+ */
+#define _LOG_RATELIMIT_CORE(_level, _rate_ms, _skipped_msg, _log_func, ...)	\
```

Comment:
Do you need to pass the `_skipped_msg`? It's the same now for all calling macros.